### ISSUE TITLE
Terraform instance profile for ASG launch config fails

### ISSUE THREAD

**Issue by anosulchik** [hashicorp/terraform#6136]
Hi guys,

We discovered the following issue with instance profile for ec2 instance in autoscaling group. During first run when launch configuration is being created along with instance profile I get the following error:
- aws_instance.w6-instance: Error launching source instance: InvalidParameterValue: IAM Instance Profile "arn:aws:iam::189366050916:instance-profile/w6-profile-tolik" 

During the next 'terraform apply' the error disappears. Looks like terraform tries to create launch configuration with instance profile that is not yet associated with IAM role.

```
resource "aws_launch_configuration" "w6-lc" {
  name_prefix = "w6-lc"
  iam_instance_profile = "${aws_iam_instance_profile.w6-instance-profile.id}"
  security_groups = [ "${aws_security_group.w6-sg.id}" ]

  depends_on = ["aws_iam_instance_profile.w6-instance-profile"]

  image_id = "ami-cb2305a1"
  instance_type = "t2.micro"
  lifecycle { create_before_destroy = true }
}
```

```
resource "aws_iam_instance_profile" "w6-instance-profile" {
  name = "w6-profile"
  roles = ["${aws_iam_role.w6-role.name}"]
}
```

P.S. terraform 0.6.14

**Comment by joshuaspence** [comment]
I have occasionally observed this issue also

**Comment by enisher** [comment]
Bump into this as well

**Comment by leventyalcin** [comment]
I've bumped the same issue. I've tried to create the launch configuration both `aws_iam_instance_profile.ec2_profile.id` and `aws_iam_instance_profile.ec2_profile.name` and when I try to create for the first time I'm getting the error below.

```
* aws_launch_configuration.drupal: Error creating launch configuration: ValidationError: Invalid IamInstanceProfile: ec2-responsive
	status code: 400, request id: 
```
and 2nd try of `terraform apply` is successful, interestingly.

```
Apply complete! Resources: 4 added, 0 changed, 0 destroyed.
```

**Comment by mtougeron** [comment]
FYI, this should be fixed via #5862

**Comment by JamesWoolfenden** [comment]
I am still getting this in the latest, instance profiles arent created by the time instance profiles need them.
User-Agent: terraform/0.6.15 () aws-sdk-go/1.1.14 (go1.6; windows; amd64)

**Comment by iceycake** [comment]
Still getting this error with 0.7.1

**Comment by AirbornePorcine** [comment]
Getting this error frequently with 0.7.4 as well.

**Comment by leventyalcin** [comment]
Well,

The main reason for this is Terraform only creates the resource and waits for the answer from API. AWS API returns "OK" once the IAM policy created, but creation of the policy does not mean it is ready to use. Any kind of change on policies takes time to be ready-to-use. 

http://docs.aws.amazon.com/IAM/latest/UserGuide/troubleshoot_general.html#troubleshoot_general_eventual-consistency

I made a workaround for this. I've separated the creation of the IAM stuff. First, I'm calling IAM terraform, sleep 5, and running terraform for rest of the stuff. 

I hope this helps for all of you too
